src/flatten_to_stream.rs

```diff
@@ -0,0 +1,101 @@
+use core::mem;
+use core::marker;
+use {Async, Future, Poll};
+use stream::Stream;
+
+/// Future for the `flatten_to_stream` combinator, flattening a future-of-a-stream to get just
+/// the result of the final future as a stream.
+///
+/// This is created by the `Future::flatten_to_stream` method.
+#[must_use = "streams do nothing unless polled"]
+pub struct FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    state: FlattenToStreamState<F>
+}
+
+pub fn new<F>(f: F) -> FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    FlattenToStream {
+        state: FlattenToStreamState::Future(f)
+    }
+}
+
+/// Stream that emits single error
+enum FailedState<T, E> {
+    Error(E, marker::PhantomData<T>),
+    Eof,
+    Done,
+}
+
+impl<T, E> Stream for FailedState<T, E> {
+    type Item = T;
+    type Error = E;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match mem::replace(self, FailedState::Done) {
+            FailedState::Done => panic!("cannot poll after eof"),
+            FailedState::Eof => {
+                *self = FailedState::Done;
+                Ok(Async::Ready(None))
+            }
+            FailedState::Error(e, _) => {
+                *self = FailedState::Eof;
+                Err(e)
+            }
+        }
+    }
+}
+
+enum FlattenToStreamState<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    // future is not yet called or called and not ready
+    Future(F),
+    // future resolved to Stream
+    Stream(F::Item),
+    // future resolved to error
+    Failed(FailedState<<F::Item as Stream>::Item, <F::Item as Stream>::Error>),
+}
+
+impl<F> Stream for FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    type Item = <F::Item as Stream>::Item;
+    type Error = <F::Item as Stream>::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            self.state = match &mut self.state {
```

Comment:
I believe you can avoid the leading `&mut` here and just bind `ref mut` in the bindings below